File: 2023/09/ts/main.ts
import type { SolutionOptions } from 'src/options.ts';
import { run } from 'src/run.ts';

export const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

function difference(ary: number[], sz: number): number[] {
   for (let i = 0; i < sz; i++) ary[i] = ary[i + 1] - ary[i];
   return ary;
}

function extrapolate(ary: number[], sz: number): number {
   sz--;
   const last = ary[sz];
   if (!sz) return last;
   return extrapolate(difference(ary, sz), sz) + last;
}

function parseInput(input: string): number[][] {
   return input.split('\n').map((str) => str.split(' ').map(Number));
}

export function part1(input: string, _isTest: boolean): string {
   const parsed = parseInput(input);
   let res = 0;
   for (const history of parsed) res += extrapolate(history, history.length);
   return res.toString();
}

export function part2(input: string, _isTest: boolean): string {
   const parsed = parseInput(input);
   let res = 0;
   for (const history of parsed) res += extrapolate(history.reverse(), history.length);
   return res.toString();
}

if (import.meta.main) {
   run(Deno.args, part1, part2, options);
}

File: 2023/09/go/main.go
package main

import (
	core "advent/of/go/src/go"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

func difference(ary []int64, sz int) []int64 {
	for i := 0; i < sz; i++ {
		ary[i] = ary[i+1] - ary[i]
	}
	return ary
}

func extrapolate(ary []int64, sz int) int64 {
	sz--
	last := ary[sz]
	if sz == 0 {
		return last
	}
	return extrapolate(difference(ary, sz), sz) + last
}

func part1(input string, isTest bool) string {
	lines := strings.Split(input, "\n")
	var res int64 = 0
	for _, line := range lines {
		nums := make([]int64, 0)
		xdd := strings.Split(line, " ")

		for _, v := range xdd {
			val, _ := strconv.ParseInt(v, 10, 64)
			nums = append(nums, val)
		}

		res += extrapolate(nums, len(nums))
	}

	return strconv.FormatInt(res, 10)
}

func part2(input string, isTest bool) string {
	lines := strings.Split(input, "\n")
	var res int64 = 0
	for _, line := range lines {
		nums := make([]int64, 0)
		xdd := strings.Split(line, " ")

		for _, v := range xdd {
			val, _ := strconv.ParseInt(v, 10, 64)
			nums = append(nums, val)
		}

		for i, j := 0, len(nums)-1; i < j; i, j = i+1, j-1 {
			nums[i], nums[j] = nums[j], nums[i]
		}

		res += extrapolate(nums, len(nums))
	}

	return strconv.FormatInt(res, 10)
}
